source.go

```diff
@@ -0,0 +1,169 @@
+package main
+
+import (
+	"fmt"
+
+	"go.ytsaurus.tech/yt/go/yson"
+)
+
+type ObjectID = string
+type SourceType string
+
+const (
+	LdapSourceType  SourceType = "ldap"
+	AzureSourceType SourceType = "azure"
+)
+
+type Source interface {
+	GetUsers() ([]SourceUser, error)
+	GetGroupsWithMembers() ([]SourceGroupWithMembers, error)
+	GetSourceType() SourceType
+}
+
+type SourceUser interface {
+	GetID() ObjectID
+	GetName() string
+	GetSourceType() SourceType
+}
+
+func NewSourceUser(sourceType SourceType, attributes map[string]any) (SourceUser, error) {
+	bytes, err := yson.Marshal(attributes)
+	if err != nil {
+		return nil, err
+	}
+
+	switch sourceType {
+	case LdapSourceType:
+		var ldapUser LdapUser
+		err = yson.Unmarshal(bytes, &ldapUser)
+		if err != nil {
+			return nil, err
+		}
+		return ldapUser, nil
+	case AzureSourceType:
+		var azureUser AzureUser
+		err = yson.Unmarshal(bytes, &azureUser)
+		if err != nil {
+			return nil, err
+		}
+		return azureUser, nil
+	}
+	return nil, fmt.Errorf("unknown source type: %v", sourceType)
+}
```

Comment:
If we having several implementations of something they better be independent. I see separation of responsibility something like that:
1) ytsaurus code knows only about interacting with YTsaurus. App should build map of user values to store for ytsaurus code (currently it is not true, YtsaurusUser have Azure fields, since I don't expect app would have more than one data source).
2) ldap code knows about fetching data from ldap and possibly about serializing/deserializing it into/from yson.
3) same for azure code
4) app know about config, have main logic of diff/create/update/delete of things and declare interfaces for Source, SourceUser, SourceGroup and so on to ldap and azure packages to satisfy those interfaces.

And it seem that those better be a separate packages. I initially made one, but currently application looks bigger than expected. But I suppose splitting to packages is out of the scope of current change, but implementations and interfaces for azure and ldap should be in different places, not like now.